<?xml version="1.0" encoding="UTF-8"?>
<!-- The simulation master launch file

Use this launch file to start the complete simulation.-->
<launch>
  <!--
       General parameters
  -->
  <!-- Name of the car used. Currently, only dr_drift is supported.-->
  <arg name="car_name" default="dr_drift" />
  <!-- Name of the simulation road. The road must be generated before it can be used!
  Run `python3 -m generate_road <NAME_OF_ROAD>` to generate a road.-->
  <arg name="road" default="default_road" />
  <!-- Indicate whether the simulation should overwrite /clock. -->
  <arg name="use_sim_time" default="true" />
  <!-- Indicate whether the graphical interface should launch. -->
  <arg name="gui" default="true" />
  <!-- Indicate whether rviz is launched. -->
  <arg name="rviz" default="false" />
  <!--
       Indicate which processes are launched.
  -->
  <!-- Indicate whether the KITcar_brain is started. -->
  <arg name="include_brain" default="true" />
  <!-- Indicate whether to launch the car_state_node. -->
  <arg name="car_state" default="true" />
  <!-- Indicate whether to launch the evaluation pipeline. This includes:
       * simulation_evaluation/speaker_node
       * simulation_evaluation/state_machine_node
       * simulation_evaluation/referee_node -->
  <arg name="evaluate" default="false" />
  <!-- Indicate whether to launch the groundtruth node. -->
  <arg name="groundtruth" default="$(arg evaluate)" />
  <!-- Indicate whether to manually control Gazebo's simulation rate.
       Normally doing this is a bad idea. However, inside of Docker containers Gazebo does not update correctly.
       Setting this argument to true starts the gazebo_rate_control_node. -->
  <arg name="control_sim_rate" default="false" />
  <!--
       Vehicle Simulation
  -->
  <!-- Whether the pose or the twist of the vehicle is given to Gazebo. -->
  <arg name="link_pose" default="true" />
  <arg name="link_twist" default="true" />
  <!--
       Gazebo
  -->
  <arg name="gz_paused" default="false" />
  <arg name="gz_recording" default="false" />
  <arg name="gz_debug" default="false" />
  <arg name="gz_physics" default="ode" />
  <arg name="gz_verbose" default="false" />
  <arg name="gz_respawn" default="false" />
  <arg name="gz_use_clock_frequency" default="false" />
  <arg name="gz_pub_clock_frequency" default="100" />
  <arg name="gz_extra_args" default="" />
  <!--
       Launch processes
  -->
  <!-- Launch KITcar_brain -->
  <group if="$(arg include_brain)">
    <include file="$(find simulation_brain_link)/launch/brain.launch">
      <arg name="car_name" value="$(arg car_name)" />
    </include>
    <!-- Launch nodes to connect gazebo and kitcar brain -->
    <include file="$(find simulation_brain_link)/launch/vehicle_link.launch">
      <arg name="car_name" value="$(arg car_name)" />
      <arg name="gui" value="$(arg gui)" />
      <arg name="rviz" value="$(arg rviz)" />
      <arg name="set_pose" value="$(arg link_pose)" />
      <arg name="set_twist" value="$(arg link_twist)" />
    </include>
  </group>
  <!-- Launch car state -->
  <group if="$(arg car_state)">
    <include file="$(find gazebo_simulation)/launch/car_state_node.launch">
      <arg name="car_name" value="$(arg car_name)" />
      <arg name="rviz" value="$(arg rviz)" />
    </include>
  </group>
  <!-- Launch evaluation pipeline -->
  <group if="$(arg groundtruth)">
    <include file="$(find simulation_groundtruth)/launch/groundtruth_node.launch">
      <arg name="road" value="$(arg road)" />
      <arg name="rviz" value="$(arg rviz)" />
    </include>
  </group>
  <include if="$(arg evaluate)" file="$(find simulation_evaluation)/launch/evaluation.launch" />
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" output="screen" args="-f simulation -d  $(find gazebo_simulation)/res/simulation.rviz -s $(find gazebo_simulation)/res/kitcar_logo.png" />
  <!-- Launch Gazebo -->
  <include file="$(find gazebo_simulation)/launch/world.launch">
    <arg name="world_name" value=".$(arg road)" />
    <arg name="paused" value="$(arg gz_paused)" />
    <arg name="use_sim_time" value="$(arg use_sim_time)" />
    <arg name="gui" value="$(arg gui)" />
    <arg name="recording" value="$(arg gz_recording)" />
    <arg name="debug" value="$(arg gz_debug)" />
    <arg name="physics" default="$(arg gz_physics)" />
    <arg name="verbose" default="$(arg gz_verbose)" />
    <arg name="respawn_gazebo" default="$(arg gz_respawn)" />
    <arg name="use_clock_frequency" default="$(arg gz_use_clock_frequency)" />
    <arg name="pub_clock_frequency" default="$(arg gz_pub_clock_frequency)" />
  </include>
  <include if="$(arg control_sim_rate)" file="$(find gazebo_simulation)/launch/gazebo_rate_control_node.launch" />
</launch>


